File: src/orchestrator/api/schemas/review.py
"""Review API schemas."""

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel


class TestRunRequest(BaseModel):
    """Request to start a test run."""

    profile: str | None = None  # Reserved for future use; v1 uses routine's auto_verify commands


class TestRunResponse(BaseModel):
    """Response when a test run is started."""

    test_run_id: str
    status: str  # "running"


class TestSummary(BaseModel):
    """Summary counts from a test run."""

    total: int
    passed: int
    failed: int
    skipped: int


class TestRunResult(BaseModel):
    """Full result of a completed (or in-progress) test run."""

    test_run_id: str
    status: str  # "running" | "passed" | "failed" | "error"
    summary: TestSummary | None = None
    log_output: str
    duration_ms: int | None = None
    started_at: datetime
    completed_at: datetime | None = None


class DiffResponse(BaseModel):
    """Response containing diff text for a branch or file."""

    diff: str
    scope: str
    file_path: str | None = None


class DiffFileEntry(BaseModel):
    """A single file entry in a diff file listing."""

    path: str
    status: str
    additions: int
    deletions: int


class CommitEntry(BaseModel):
    """A single commit entry in a commit history listing."""

    sha: str
    short_sha: str
    message: str
    author: str
    timestamp: datetime


class LineRange(BaseModel):
    """A range of line numbers (inclusive)."""

    start: int
    end: int


class FilePrune(BaseModel):
    """Prune selection for a single file."""

    path: str
    mode: Literal["file", "hunk", "line"]
    hunks: list[int] | None = None
    lines: list[LineRange] | None = None


class PruneSelection(BaseModel):
    """A set of prune selections across files."""

    files: list[FilePrune]
    scope: Literal["aggregate", "commit", "task"]


class PrunePreviewResponse(BaseModel):
    """Preview of what a prune operation would produce."""

    resulting_diff: str
    files_affected: int
    hunks_removed: int
    lines_removed: int


class PruneApplyResponse(BaseModel):
    """Result of applying a prune operation."""

    commit_sha: str
    files_affected: int
    hunks_removed: int
    lines_removed: int
    event_id: str


class ConflictBlock(BaseModel):
    """A single conflict block within a file."""

    index: int
    ours_content: str
    theirs_content: str
    base_content: str | None = None


class ConflictFile(BaseModel):
    """A file with unresolved merge conflicts."""

    path: str
    status: str  # "unresolved" | "resolved"
    block_count: int
    blocks: list[ConflictBlock]


class BlockResolution(BaseModel):
    """Resolution for a single conflict block."""

    block_index: int
    choice: str  # "ours" | "theirs" | "manual"
    manual_content: str | None = None


class ConflictResolutionRequest(BaseModel):
    """Request to resolve conflicts in a file."""

    resolutions: list[BlockResolution]


class ConflictResolutionResponse(BaseModel):
    """Response after resolving conflicts in a file."""

    path: str
    status: str  # "resolved"
    remaining_conflicts: int


class BackMergeResponse(BaseModel):
    """Response from a back merge operation."""

    status: str  # "clean" | "conflicts"
    merge_commit_sha: str | None = None
    conflict_files: list[str] = []
    conflict_count: int = 0


class Gate(BaseModel):
    """A single readiness gate with pass/fail/pending status."""

    name: str
    status: str  # "pass" | "fail" | "pending"
    description: str


class MergeReadiness(BaseModel):
    """Aggregate merge readiness computed from all gates."""

    ready: bool
    gates: list[Gate]


class RevertFileRequest(BaseModel):
    """Request to revert a single file to base-branch state."""

    file_path: str


class AgentResolveConflictsRequest(BaseModel):
    """Request for agent-based conflict resolution."""

    agent_type: str | None = None
    agent_config: dict[str, Any] | None = None

File: ui/src/types/review.ts
/**
 * TypeScript types matching backend review API schemas.
 * See: src/orchestrator/api/schemas/review.py
 */

export interface DiffResponse {
  diff: string;
  scope: string;
  file_path: string | null;
}

export interface DiffFileEntry {
  path: string;
  status: "added" | "modified" | "deleted" | "renamed";
  additions: number;
  deletions: number;
}

export interface CommitEntry {
  sha: string;
  short_sha: string;
  message: string;
  author: string;
  timestamp: string;
}

export interface LineRange {
  start: number;
  end: number;
}

export interface FilePrune {
  path: string;
  mode: 'file' | 'hunk' | 'line';
  hunks: number[] | null;
  lines: LineRange[] | null;
}

export interface PruneSelection {
  files: FilePrune[];
  scope: string;
}

export interface PrunePreviewResponse {
  resulting_diff: string;
  files_affected: number;
  hunks_removed: number;
  lines_removed: number;
}

export interface PruneApplyResponse {
  commit_sha: string;
  files_affected: number;
  hunks_removed: number;
  lines_removed: number;
  event_id: string;
}

export interface TestRunResponse {
  test_run_id: string;
  status: string; // "running"
}

export interface TestSummary {
  total: number;
  passed: number;
  failed: number;
  skipped: number;
}

export interface TestRunResult {
  test_run_id: string;
  status: 'running' | 'passed' | 'failed' | 'error';
  summary: TestSummary | null;
  log_output: string;
  duration_ms: number | null;
  started_at: string;
  completed_at: string | null;
}

export interface AgentJobResponse {
  job_id: string;
  status: string;
}

export interface ConflictBlock {
  index: number;
  ours_content: string;
  theirs_content: string;
  base_content: string | null;
}

export interface ConflictFile {
  path: string;
  status: 'unresolved' | 'resolved';
  block_count: number;
  blocks: ConflictBlock[];
}

export interface BlockResolution {
  block_index: number;
  choice: 'ours' | 'theirs' | 'manual';
  manual_content?: string | null;
}

export interface ConflictResolutionResponse {
  path: string;
  status: string;
  remaining_conflicts: number;
}

export interface BackMergeResponse {
  status: 'clean' | 'conflicts';
  merge_commit_sha: string | null;
  conflict_files: string[];
  conflict_count: number;
}

export interface RevertBackMergeResponse {
  reverted_commit: string;
  new_head: string;
}

export interface Gate {
  name: string;
  status: 'pass' | 'fail' | 'pending';
  description: string;
}

export interface MergeReadiness {
  ready: boolean;
  gates: Gate[];
}

export interface FinalMergeBackResponse {
  merge_commit: string;
  strategy: string;
  message: string;
}
